lib/vdsm/virt/thinp.py

```diff
@@ -553,24 +592,23 @@ def _extend_volume_completed(self, volInfo):
                 with clock.run("refresh-destination-volume"):
                     self._vm.refresh_destination_volume(volInfo)
 
-            with clock.run("refresh-volume"):
-                self._vm.refresh_volume(volInfo)
-
-            # Check if the extension succeeded.  On failure an exception is
-            # raised.
-            # TODO: Report failure to the engine.
-            volSize = self._verify_volume_extension(volInfo)
+            # After a volume was refreshed, the montior thread may wake up and
+            # trigger unwanted extend to the next chunk size. Lock the monitor
+            # during refresh until we set a new threshold.
+            with self._lock:
```

Comment:
This lock looks somewhat scary.  It's difficult to trace here what's locked when and this one is not timed so there is a risk of deadlock.  Well, the only place where this method is used is the callback in _extend_volume and this is asynchronous so we should be safe, right?

Another problem may be that the lock will block certain operations on any of the drives.  This shouldn't cause harm because we process the extension requests sequentially, right?

Is there anything performed under the lock that could take significant time?  Could it cause some part of the whole mechanism to run much less smoothly than possible?

And finally, any risk of competing for the lock from multiple threads that could cause starvation (related to the previous paragraph)?  Which threads are involved?  Only monitoring and mailbox?  libvirt events don't invoke the lock, right?  Anything else?

Just thinking aloud -- it seems to me that it should work well but it's better to double-check.

Reply:
Yes, taking the lock without a timeout is risky.

The operation that can block are:
- refresh volume (lvchange --refresh)
- getting volume size (fsutils.size())
- setting block threshold in libvirt

The first 2 access storage so they can block up to 80 seconds on properly configured host. On block storage we configure multipath to fail after 80 seconds. But a user can override this configuration and it this case accessing storage can block for long time. But if accessing storage blocks, the VM will be unresponsive and the guest will be block on storage.

When the lock is taken, extending or refreshing other drives will be blocked, which is not good, in case you have drives on multiple storage domains and only one of them blocks. Ideally we would lock only the relevant drive instead of the entire monitor.

I think we can make this safer by using the same self._timed_lock() with a larger
timeout, maybe 20 seconds? If we time out, the monitor will retry the extend every 2 seconds so we should get a new completion event soon after the timeout.

Of course if we time out, and the refresh completed after the timeout, the next monitor cycle may try to extend again to a larger size - what this lock is trying to avoid.

Reply:
This PR is little to big, I plan to split it to 2 parts, and move the locking patches to the second part, so we can merge the simpler and safe changes first.

Reply:
> This PR is little to big, I plan to split it to 2 parts, and move the locking patches to the second part, so we can merge the simpler and safe changes first.

OK.  Reviewing updates to large PRs is difficult -- I don't have a tool that would show me which commits where changed and how.  Unless the PR update is small and not containing unrelated changes due to rebase, there is nothing better to do than to review all the patches again, which is a lot of work.  So better to avoid large PRs.  (Yes, maintaining PRs depending on other PRs is another hassle then, but probably still less trouble.)

Reply:
> > I think we can make this safer by using the same self._timed_lock() with a larger
> > timeout, maybe 20 seconds? If we time out, the monitor will retry the extend every 2 seconds so we should get a new completion event soon after the timeout.
> 
> So what would exactly happen if the whole block protected by the lock was skipped? I thought this method could be called again only if a new extension was attempted.

If we time out locking before refresh, the the refresh will not happen. This means the size of the logical volume remain the same, so the VM is likey to pause.

Since the drive is still marked as EXCEEDED, the system will try to send more extend request, which will succeed (doing nothing, since the logical volume was already extended). The next extend completion will retry to the lock and do the refresh.

This does not change the way the system is already working, just prevent the race after refreshing a logical volume.

> > Of course if we time out, and the refresh completed after the timeout, the next monitor cycle may try to extend again to a larger size - what this lock is trying to avoid.
> 
> Which refresh? The one here is protected by the lock and won't be performed at all if the lock is not taken.

Actually if the lock times out no refresh is started, so timeout on this lock is safe.